Added support to mark HystrixBadRequestException as metrics (in)transient

matelang opened the pull request:
See issue for proposition: https://github.com/Netflix/Hystrix/issues/1807

mattnelson commented:
`HystrixBusinessException` sounds like corporate speak and doesn't clearly indicate what is actually wrong. `HystrixBadRequestException` already covers this. I would propose to instead introduce a property that includes `HystrixBadRequestException` with the statistics. That would also allow all current consumers to opt into this feature with only a config change.

matelang commented:
Hi @mattnelson ,

Thanks for taking the time to look into this.

I am happy to change the name of the exception if it sounds too corporate, unfortunately I can not think of a better name to express the fact that we'd like to throw an exception that indicates that the called method was indeed executed (opposed to the `HystrixBadRequestException` case which, if I understand correctly should translate to something which could not be executed and should be completely ignored from statistics).

Imagine the following example scenario:

```java
public class UserNotFoundException extends HystrixBusinessException{
 //ctors omitted for brevity
}

public interface UserService {
  User findUser(String userName) throws UserNotFoundException;  
}

public class RemoteUserService implements UserService{
  @Override
  public User findUser(String userName){
    ResponseEntity<User> resp = restTemplate.get(...); // get the user
    if (resp.getStatusCode() == 404){   
      throw new UserNotFoundException();
    }

    return resp.getBody();
  }
}
```

I'd expect to treat the user not being found case from the previous example as an absolute success of the operation, because the implementation fulfilled its purpose.
If the above method `findUser` is called from a Hystrix command I'd expect the command to be marked as successful.

I understood that the `HystrixBadRequestException` is more suitable for situations where the operation was skipped or not performed. As an example: for a web service we could skip the actual call if we are unable to serialize an object to be sent to JSON. That means we did not actually call the service, so we should not touch any of the metrics nor the state of the circuit breaker.

I am not sure whether the 2 cases: `HystrixBadRequestException` and `HystrixBusinessException` should be treated as one.

The proposed solution is also non-intrusive. Clients who do not opt in for the BusinessException capability will not see any change in behavior.

mattnelson commented:
> I'd expect to treat the user not being found case from the previous example as an absolute success of the operation, because the implementation fulfilled its purpose.
> If the above method findUser is called from a Hystrix command I'd expect the command to be marked as successful.

That still complies with what a `HystrixBadRequestException` was intended for. This may be straddling the line of semantics, but a 404 is a result of the client providing bad input, thus a bad request. Thats why I'm pushing for using the existing exception type, and introducing a toggle to allow these requests to count for success metrics.

https://github.com/Netflix/Hystrix/blob/7f5a0afc23aa5ff82320560a04d4c81a45efd67c/hystrix-core/src/main/java/com/netflix/hystrix/exception/HystrixBadRequestException.java#L20-L27

matelang commented:
It is indeed semantics.

At the end of the day I'm happy to go with any solution that fulfills the end goal here.
Were do you think the toggle would be most suitable in? I'm happy to modify the PR accordingly.

Cheers.

matelang commented:
Hi @mattnelson.

I've changed the code according to your requests, by introducing `metricsTransient` field on the BadRequestException and treating it accordingly in AbstractCommand.

I hope it's okay now.

arnoldtempfli commented:
Any update on this?

matelang commented:
Hi @mattnelson ,

do you want me to do anything else on this PR?
If not can you please merge it?

mattnelson commented on hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java:
```diff
@@ -602,15 +602,23 @@ public void call() {
                     return handleTimeoutViaFallback();
                 } else if (t instanceof HystrixBadRequestException) {
-                    return handleBadRequestByEmittingError(e);
+                    HystrixBadRequestException hbre = (HystrixBadRequestException) e;
```
Should be `t` instead of `e` since the else block does an instance of check on `t`.

mattnelson commented on hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java:
```diff
@@ -602,15 +602,23 @@ public void call() {
+                        circuitBreaker.markSuccess();
+                    }
+
+                    return handleBadRequestByEmittingError(hbre);
```
Shouldn't this be `e` to remain passive?

matelang commented:
Hi @mattnelson ,

Thanks for your guidance in implementing this.
I've addressed your comments.

Do you know who could I contact for having this merged in?

Cheers,

matelang commented:
Can somebody please take a look into this? :cry: 
This is causing us issues in production.

matelang commented:
Can somebody please take a look? :( 
Still causing production issues. I'd really like to avoid building our own Hystrix :cry:

takZhang commented:
I've never used .  so sorry

gvespucci commented:
Hi @matelang did you receive any feedback on this? Looks like you gave up at the end... 🥺

matelang commented:
@gvespucci it seems that nobody is taking care of this repo anymore.
